### PR TITLE
Adding workload manifest offline caching

### PR DESCRIPTION
**Pull request by sfoslund**
Finishes offline caching for workload install, prepares for workload update offline caching

### CODE REVIEW

File: src/Tests/dotnet-workload-install.Tests/MockWorkloadManifestUpdater.cs [hunk 1]
```diff
@@ -2,23 +2,29 @@
...
 using Microsoft.DotNet.Workloads.Workload.Install;
+using Microsoft.Extensions.EnvironmentAbstractions;
 
 namespace Microsoft.DotNet.Cli.Workload.Install.Tests
 {
     internal class MockWorkloadManifestUpdater : IWorkloadManifestUpdater
     {
         public int UpdateAdvertisingManifestsCallCount = 0;
         public int CalculateManifestUpdatesCallCount = 0;
+        public int DownloadManifestPackagesCallCount = 0;
```
Comment (wli3) [inline src/Tests/dotnet-workload-install.Tests/MockWorkloadManifestUpdater.cs:16]: nit: Good posts on mock and some solution to reduce them http://arlobelshee.com/tag/no-mocks/

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs [hunk 2]
```diff
@@ -79,22 +131,37 @@ private IEnumerable<ManifestId> GetInstalledManifestIds()
...
+                    }
                 }
-                catch (NuGetPackageNotFoundException)
+                else
                 {
-                    _reporter.WriteLine(string.Format(LocalizableStrings.AdManifestPackageDoesNotExist, manifestId));
+                    packagePath = Directory.GetFiles(offlineCache.Value.Value)
+                        .Where(path => path.EndsWith(".nupkg"))
+                        .Where(path => Path.GetFileName(path).StartsWith(GetManifestPackageId(_sdkFeatureBand, manifestId).ToString()))
+                        .FirstOrDefault();
```
Comment (wli3) [inline src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs:158]: Should we get the highest? Or throw on there are more than one valid manifest?

File: src/Tests/dotnet-workload-install.Tests/GivenDotnetWorkloadInstall.cs [hunk 3]
```diff
@@ -126,16 +126,17 @@ public void GivenWorkloadInstallItCanUpdateInstalledManifests()
         [Fact]
         public void GivenWorkloadInstallItCanDownloadToOfflineCache()
```
Comment (wli3) [inline src/Tests/dotnet-workload-install.Tests/GivenDotnetWorkloadInstall.cs:127]: Could you add a test to make sure it is the manifest in cache get used(to find the packs)? and later installed (placed on disk)

File: src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs [hunk 4]
```diff
@@ -216,31 +220,73 @@ public void InstallWorkloads(IEnumerable<WorkloadId> workloadIds, bool skipManif
...
             var packageUrls = new List<string>();
+            DirectoryPath? tempPath = null;
 
-            if (_workloadInstaller.GetInstallationUnit().Equals(InstallationUnit.Packs))
+            try
             {
-                var installer = _workloadInstaller.GetPackInstaller();
+                if (!skipManifestUpdate)
+                {
+                    // Add package urls to print
```
Comment (wli3) [inline src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:232]: nit: whenever you see comments for "steps", you could extract a method here and name it AddPackagesUrlsToPrint

Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:232]: So is UseManifestsToResolvePacks

**Review by wli3 — APPROVED** [review]

**Comment by sfoslund** [comment]
@marcpopMSFT do we need this in preview 5?

**Comment by marcpopMSFT** [comment]
Let's chat as a group in the workload meeting tomorrow.  There's still time for preview 5.

**Pull request merged**